[Impeller] Add rounded superellipse

dkwingsmt opened the pull request:
Support rounded superellipse.

Part of https://github.com/flutter/flutter/issues/139321 and https://github.com/flutter/flutter/issues/13914, also related to https://github.com/flutter/flutter/issues/91523.

### Open questions

* Alternative names:
  * Round**ed**Superellipse
  * Squircle
  * ContinuousBorderRectangle (or something like this...)
  * I chose rounded superellipse because this name, albeit its length, precisely describe this shape. "Squircle" is not strictly defined but generally refers any shape intermediate between a rectangle and a circle.
* Alternative definition for `corner_radius`:
  * Currently the `corner_radius` corresponds to SwiftUI parameters. To make the shape definition more generalized, we can instead define the `corner_radius` to be the radius of the corner circles, and make the framework do a look up table.
  * The down side is, not only the work to re-calculate the table, but also that it doesn't completely eliminates the relationship with SwiftUI, since currently the degree of the superellipse (`n`) is also mapped from the SwiftUI `cornerRadius`, which is not necessary for the shape per se.
  * To some extent it boils down to the question of whether we'd like this shape to support anything beyond SwiftUI.

### Demo

https://github.com/user-attachments/assets/806ac0e9-d62f-4b04-ab6a-83436a11f6f3

Low ratio: (900, 900, 445)

<img width="520" alt="image" src="https://github.com/user-attachments/assets/54087467-85cd-4021-91cc-a948866ab5a8">

Mid ratio: (900, 650, 180)

<img width="508" alt="image" src="https://github.com/user-attachments/assets/460a4927-0396-462b-948d-0846a781c92c">

High ratio: (900, 650, 17)

<img width="490" alt="image" src="https://github.com/user-attachments/assets/8d7f625d-8a3b-4aba-b3f9-f292b874b606">


## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+   *  C ⟋  ↙
+   */
+
+  const Point s_to_e = end - start;
```
Please drop `const` for all of these locals.

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+
+  // Points without applying `flip` and `center`.
+  std::vector<Point> points;
+  points.reserve(21);
```
What is this number from?

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+  {
+    const Scalar target_slope = yJ / xJ;
+    Scalar angle = 0;
+    while (true) {
+      const Scalar x = a * pow(abs(sinf(angle)), 2 / n);
+      const Scalar y = a * pow(abs(cosf(angle)), 2 / n);
+      if (y <= target_slope * x) {
+        break;
+      }
+      points.emplace_back(x + s, y + s);
+      angle += kAngleStep;
+    }
```
consider `do/while` instead.

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+  // Draw the first quadrant of the shape and store in `points`. It will be
+  // mirrored to other quadrants later.
+  std::vector<Point> points;
+  points.reserve(41);
```
same here, what is this number from?

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+  // The iteration order is reversed so that the trianges are continuous from
+  // quadrant to quadrant.
+  std::vector<Point> geometry;
+  geometry.reserve(1 + 4 * points.size());
```
Once you know the amount of storage needed, its better to write directly into the host buffer as it avoids an extra copy.

Example:

```

 BufferView point_buffer = host_buffer.Emplace(
      nullptr, sizeof(Point) * point_count, alignof(Point));
Point* point_data =  reinterpret_cast<Point*>(point_buffer.GetBuffer()->OnGetContents() +
                                   point_buffer.GetRange().offset);
/// write to point_data                                   
 ```

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+  }
+  geometry.push_back(center + points[0]);
+
+  std::vector<uint16_t> indices;
```
Same here

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+  std::vector<Point> points;
+  points.reserve(41);
+
+  DrawOctantSquareLikeSquircle(points, size.width, corner_radius_, Point{0, -c},
```
The tessellator has a large point area already allocated for temp workloads (ignore name): https://github.com/flutter/engine/blob/main/impeller/tessellator/tessellator.h#L310 . This has 4096 slots of storage. If you can handle overflows by writing to a temp buffer, then its going to be much faster to write there instead of to a newly created vector.

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+
+bool RoundSuperellipseGeometry::CoversArea(const Matrix& transform,
+                                           const Rect& rect) const {
+  return false;
```
I'd imagine computing a conservative inner rect for the rounded superellipse should be possible, no?

jonahwilliams commented on impeller/entity/geometry/round_superellipse_geometry.cc:
```diff
@@ -0,0 +1,306 @@
+
+  std::vector<uint16_t> indices;
+  indices.reserve(geometry.size() * 3);
+  for (auto i = 2u; i < geometry.size(); i++) {
```
Are you generating a triangle fan structure? if so you can use the triangle fan geometry type instead and skip the index buffer (where supported).

jonahwilliams reviewed (commented):
Really excited to see this work coming along!

jonahwilliams commented:
I changed canvas.drawRect to use this geometry and then ran an app that draws a few dozen rectangles to get this profile. It looks like the vast majority of overhead is from temporary allocation. That should disappear if you use the preallocated arenas.

Also kAngleStep may be too small for small rectangles. I'm getting ~182 points for a 20x20 squircle. You may want to eyeball some numbers for this based off of the rectangle size.

![image](https://github.com/user-attachments/assets/1455c36b-4e92-436e-9693-91882ef07fdd)

mit-mit commented:
I suggest we simplify the name to just "Superellipse". That matches Peit Heins coinage of the term in 1965, where he used the "super"-prefix to denote those curves where `n > 2`:
https://piethein.com/superellipse/
https://en.wikipedia.org/wiki/Superellipse

dkwingsmt commented:
@mit-mit That's the definition I adopted as well. However, my shape is a superellipse of this definition combined with 4 circles at the corner.

mit-mit commented:
Sorry, I'm confused, don't all superellipses have rounded corners? And isn't a squircle just a superellipse with n=4? https://news.ycombinator.com/item?id=37871982

dkwingsmt commented:
> isn't a squircle just a superellipse with n=4?

Well, to my understanding, squircles are a general term that describe any shape intermediate between a rectangle and an oval. There can be infinitely many kinds of squircles. https://en.wikipedia.org/wiki/Squircle

According to [Jonah's previous PR](https://github.com/flutter/engine/pull/54562), superellipse with n=4 is called rectellipse. 

> don't all superellipses have rounded corners?

You're right, but we're looking for a very special shape, _the_ shape used by Apple in its designs (which I call _the Apple squircle_). I've found (based on prior arts) that the Apple squircle looks very much like a superellipse, but with the rounded corners more dull. By replacing the tip of the rounded corners with appropriate circular arcs, the resulting shape happens to be almost, if not exactly, identical to the Apple squircles.

I have a technical report in store, where I'll describe my research process and the math behind. My apology that I've only shared the tech report with a few folks, mostly because the PR is still experimental. I'll publish the tech report soon. But feel free to ask any questions!

mit-mit commented:
Thanks for all the details, glad to see so much care being put into this!

dkwingsmt commented:
@jonahwilliams I've addressed all your comments. Notably:

* The result is now always arranged in a triangle strip (without indices), which should be the most performant way with max compatibility.
* Intermediate data are now stored in the tessellator cache. In fact, the source file doesn't import `<vector>` anymore.
* The angular step is now calculated dynamically based on dimension. This has reduced the number of points drastically: ~55 for large rectangles with small radius; ~35 for really small rectangles; ~150 for a large, full circle (which has the most curves and thus requiring most points).

PTAL, thanks!

rydmike commented:
This is awesome, can't wait for it to become available as shape in Flutter. When it does I will update this and it to the comparison study https://github.com/rydmike/squircle_study
